Add bitfields to metadata

mat-1 opened the pull request:
Adds a "bitfields" parameter in the metadata which includes method, value (bitmask), and (sometimes) class
![screenshot of abstract_entity bitfields](https://user-images.githubusercontent.com/27899617/108642692-6a154180-746c-11eb-97e3-95044f4bd628.png)

Closes #17 

(i hope i rebased correctly this time lol)

Pokechu22 committed:
Don't choke if light_setter is None

That currently happens in 20w12a, since the light setter now is a lambda that gets the block state.  Currently, this means there is no lighting information in 20w12a, but I'll fix that soon.

Pokechu22 committed:
Call additional methods within the block list class as needed

This fixes the remaining errors (due to current_block being None).  Some hardness information is missing now for stairs and slabs, but I don't think that's easily solvable.

Pokechu22 committed:
Reimplement light levels for 20w12a

This fails in a few more cases than just sea pickles, but most of them should be fixable.

Pokechu22 committed:
Stop printing full backtrace when light data lambda fails

Since this occurs naturally for 2 blocks, the backtrace is just noise.

Pokechu22 committed:
Identify the pose metadata serializer

Pokechu22 committed:
Improve identification of BlockState

The previous heuristic broke in 20w21a.

Pokechu22 committed:
Identify VillagerData

Pokechu22 committed:
Identify Particle serializer

Now the serializer identification code is much less hacky.

Pokechu22 committed:
Fix KeyError: 'particle' in older versions

Pokechu22 committed:
Fix identification of entity list in 20w30a

Refactoring in that snapshot moved the disconnect.lost string constant in NHPC into a static initializer, which put it at the end of the constant list.  NHPC also has the Skipping Entity with id constant, which means that it ended up matching as the entity list instead (since that was now before disconnect.lost).  Then the actual entity list also matched, which threw an exception.

The fix is to make Skipping Entity with id a lower priority constant.

nickelpro committed:
Add particletypes topping

mat-1 committed:
Add bitfield extraction

mat-1 committed:
fix some bitfields missing for abstract_entity

mat-1 committed:
re-add comment i accidentally removed

Pokechu22 commented:
Code looks good to me and seems to run correctly on all versions where entity metadata works (I tested 1.9, 1.12.2, and 21w07a).  I think `mask` might be a clearer name than `value` though.  There are also a few minor spelling and whitespace errors, but I can fix all of those up afterwards.

The rebase doesn't seem to have gone perfectly, but I can do a squash merge which should avoid causing issues in the main repo (it'll also only create one commit, so the extra fixup commits won't show up).

mat-1 committed:
value -> mask

mat-1 committed:
remove random trailing space

mat-1 committed:
change single quote to double quote for consistency

mat-1 commented:
> 
> 
> Code looks good to me and seems to run correctly on all versions where entity metadata works (I tested 1.9, 1.12.2, and 21w07a). I think `mask` might be a clearer name than `value` though. There are also a few minor spelling and whitespace errors, but I can fix all of those up afterwards.
> 
> The rebase doesn't seem to have gone perfectly, but I can do a squash merge which should avoid causing issues in the main repo (it'll also only create one commit, so the extra fixup commits won't show up).

Ok, I changed it from `value` to `mask` and fixed some of those issues

Pokechu22 commented on burger/toppings/entitymetadata.py:
```diff
@@ -231,8 +236,76 @@ def on_new(self, ins, const):
+                            is_getflag_method = has_correct_arguments and calling_method == get_flag_method
+                            is_shared_getflag_method = has_correct_arguments and calling_method == shared_get_flag_method
+
+                            # if its a shared flag, update the bitfields_by_class for abstract_entity
```
Since you seem to have addressed everything else, this should be <q>it's</q>.

mat-1 committed:
its -> it's

Pokechu22 commented:
Merged, thanks!  This data will be included in the next snapshot burger runs on (probably 21w08a).  I'll eventually re-run burger on _all_ the snapshots, but that won't happen for a while.

I did just now think of one other thing that might be useful: getting the actual metadata field associated with the bitfield.  But currently, no entity has more than one bitfield, so I don't think that's too important; if that ever changes I don't think it'd be too hard to add.

mat-1 commented:
I'm like 90% sure the first byte is always the bitfield

Pokechu22 commented:
It's the last byte for [pandas](https://wiki.vg/Entity_metadata#Panda).
